nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/main/java/org/nuxeo/ecm/platform/comment/impl/CommentsMigrator.java

```diff
@@ -59,64 +74,166 @@
 
     protected static final int BATCH_SIZE = 50;
 
+    /**
+     * @since 11.1.
+     */
+    public static final String GET_COMMENTS_FOLDERS_QUERY = String.format(
+            "SELECT %s FROM Document WHERE %s = '%s' AND %s ='%s'", ECM_UUID, ECM_NAME, COMMENTS_DIRECTORY,
+            ECM_PRIMARYTYPE, HIDDEN_FOLDER_TYPE);
+
     @Override
     protected String probeSession(CoreSession session) {
-        CommentService commentComponent =  Framework.getService(CommentService.class);
+        CommentService commentComponent = Framework.getService(CommentService.class);
         CommentServiceConfig commentServiceConfig = commentComponent.getConfig();
         if (commentServiceConfig != null) {
             Graph graph = Framework.getService(RelationManager.class).getGraph(commentServiceConfig.graphName, session);
-            if (graph.getStatements().size() > 0) {
+            if (!graph.getStatements().isEmpty()) {
                 return MIGRATION_STATE_RELATION;
             }
         }
-        return MIGRATION_STATE_PROPERTY;
+        // If not in relation, check if there are still property comments under hidden Comments folder(s)
+        // Do a new query one to check the step and the second the make the migration (do all comments)
+        if (hasUnsecuredComments(session)) {
+            return MIGRATION_STATE_PROPERTY;
+        }
+        // Else, comments are already secured
+        return MIGRATION_STATE_SECURED;
     }
 
-    @Override
-    protected void migrateSession(CoreSession session) {
+    /**
+     * @since 11.1
+     */
+    protected List<DocumentModel> getUnsecuredComments(CoreSession session) {
+        List<String> parentIds = getCommentFolders(session);
+        if (parentIds.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return session.query(buildAllCommentsQuery(parentIds, COMMENT_DOC_TYPE));
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected boolean hasUnsecuredComments(CoreSession session) {
+        List<String> folderIds = getCommentFolders(session);
+        return folderIds.stream().anyMatch(folderId -> session.hasChildren(new IdRef(folderId)));
```

Comment:
Do we need it? We just retrieved all ids, so they suppose to exist no?

Reply:
Yes, we took the hypothesis that we can have a customers which can have some comment hidden folders that have been emptied.